pkg/kernel/nfqueue.go

```diff
@@ -131,139 +152,315 @@ func (nfq *NFQueue) Update(protocols []string, sourceIPs []string, destinationIP
 	if err != nil {
 		return err
 	}
-	err = execCommand(fmt.Sprintf("nft add set inet %s %s { type ipv4_addr\\; flags interval\\; }", tableName, nfq.getFullSetName(ipv4DestinationIPSet)))
-	if err != nil {
-		return err
+	nfq.ipv4SourceSet = &nftables.Set{
+		Table:    nfq.table,
+		Name:     nfq.getFullSetName(ipv4SourceIPSetName),
+		Interval: true,
+		KeyType:  nftables.TypeIPAddr,
 	}
-	err = execCommand(fmt.Sprintf("nft add set inet %s %s { type ipv4_addr\\; flags interval \\; }", tableName, nfq.getFullSetName(ipv4SourceIPSet)))
+	err = conn.AddSet(nfq.ipv4SourceSet, []nftables.SetElement{})
 	if err != nil {
 		return err
 	}
-	err = execCommand(fmt.Sprintf("nft add set inet %s %s { type ipv6_addr\\; flags interval \\; }", tableName, nfq.getFullSetName(ipv6DestinationIPSet)))
-	if err != nil {
-		return err
+	nfq.ipv6DestinationSet = &nftables.Set{
+		Table:    nfq.table,
+		Name:     nfq.getFullSetName(ipv6DestinationIPSetName),
+		Interval: true,
+		KeyType:  nftables.TypeIP6Addr,
 	}
-	err = execCommand(fmt.Sprintf("nft add set inet %s %s { type ipv6_addr\\; flags interval \\; }", tableName, nfq.getFullSetName(ipv6SourceIPSet)))
+	err = conn.AddSet(nfq.ipv6DestinationSet, []nftables.SetElement{})
 	if err != nil {
 		return err
 	}
-	err = execCommand(fmt.Sprintf("nft add set inet %s %s { type inet_service\\; flags interval \\; }", tableName, nfq.getFullSetName(destinationPortSet)))
-	if err != nil {
-		return err
+	nfq.ipv6SourceSet = &nftables.Set{
+		Table:    nfq.table,
+		Name:     nfq.getFullSetName(ipv6SourceIPSetName),
+		Interval: true,
+		KeyType:  nftables.TypeIP6Addr,
 	}
-	err = execCommand(fmt.Sprintf("nft add set inet %s %s { type inet_service\\; flags interval \\; }", tableName, nfq.getFullSetName(sourcePortSet)))
+	err = conn.AddSet(nfq.ipv6SourceSet, []nftables.SetElement{})
 	if err != nil {
 		return err
 	}
-	err = execCommand(fmt.Sprintf("nft add chain inet %s %s { type filter hook prerouting priority %d \\; }", tableName, nfq.name, nfq.priority))
+	nfq.protocolSet = &nftables.Set{
+		Table:   nfq.table,
+		Name:    nfq.getFullSetName(protocolSetName),
+		KeyType: nftables.TypeInetProto,
+	}
+	err = conn.AddSet(nfq.protocolSet, []nftables.SetElement{})
 	if err != nil {
 		return err
 	}
-	return nil
+	chain := &nftables.Chain{
+		Name:     nfq.name,
+		Table:    nfq.table,
+		Type:     nftables.ChainTypeFilter,
+		Hooknum:  nftables.ChainHookPrerouting,
+		Priority: nftables.ChainPriority(nfq.priority),
+	}
+	nfq.chain = conn.AddChain(chain)
+	return conn.Flush()
+}
+
+func (nfq *NFQueue) configureRules() error {
+	conn := &nftables.Conn{}
+	// nft add rule inet meridio-nfqlb flow-a meta l4proto @flow-a-protocols ip saddr @flow-a-saddrs-v4 ip daddr @flow-a-daddrs-v4 th dport @flow-a-dports th sport @flow-a-sports counter queue num 1
+	ipv4Rule := &nftables.Rule{
+		Table: nfq.table,
+		Chain: nfq.chain,
+		Exprs: []expr.Any{
+			// [ meta load l4proto => reg 1 ]
+			&expr.Meta{
```

Comment:
"Multiple expressions are linearly evaluated from left to right: if the first expression matches, then the next expression is evaluated and so on."
Probably performance wise it'd be better checking first if the packet is IPv4 (instead the L4 protocol type)

Reply:
Do you mean inverting:
```
 meta load l4proto => reg 1 ]
[ lookup reg 1 set flow-a-protocols ]
```
with:
```
[ meta load nfproto => reg 1 ]
[ cmp eq reg 1 0x00000002 ]
```
I just took these 2 commands:
```
nft add rule inet meridio-nfqlb flow-a meta l4proto @flow-a-protocols ip saddr @flow-a-saddrs-v4 ip daddr @flow-a-daddrs-v4 th dport @flow-a-dports th sport @flow-a-sports counter queue num 1
nft add rule inet meridio-nfqlb flow-a meta l4proto @flow-a-protocols ip6 saddr @flow-a-saddrs-v6 ip6 daddr @flow-a-daddrs-v6 th dport @flow-a-dports th sport @flow-a-sports counter queue num 1
```
and used "--debug all" which returns all expressions, and I converted it to the code.

Reply:
Yes, that what i meant, to have the IP version check first